internal/hostname/providers.go

```diff
@@ -0,0 +1,243 @@
+		return cachedHostname, cachedProvider, true
+	}
+	return cachedHostname, cachedProvider, false
+}
+
+// setCached caches the newHostname
+func setCached(now time.Time, newHostname string, newProvider string) {
+	m.Lock()
+	defer m.Unlock()
+	cachedHostname = newHostname
+	cachedAt = now
+	cachedProvider = newProvider
+}
+
+type provider struct {
+	name string
+	// Should we stop going down the list of providers if this one is successful
+	stopIfSuccessful bool
+	pf               providerFetch
+}
+
+type providerFetch func(ctx context.Context, currentHostname string) (string, error)
+
+var providerCatalog = []provider{
+	{
+		name:             "configuration",
+		stopIfSuccessful: true,
+		pf:               fromConfig,
+	},
+	{
+		name:             fargateName,
+		stopIfSuccessful: true,
+		pf:               fromFargate,
+	},
+	{
+		name:             "gce",
+		stopIfSuccessful: true,
+		pf:               fromGce,
+	},
+	{
+		name:             "azure",
+		stopIfSuccessful: true,
+		pf:               fromAzure,
+	},
+	// The following providers are coupled. Their behavior changes depending on the result of the previous provider.
+	// Therefore, 'stopIfSuccessful' is set to false.
+	{
+		name:             "fqdn",
+		stopIfSuccessful: false,
+		pf:               fromFQDN,
+	},
+	{
+		name:             "container",
+		stopIfSuccessful: false,
+		pf:               fromContainer,
+	},
+	{
+		name:             "os",
+		stopIfSuccessful: false,
+		pf:               fromOS,
+	},
+	{
+		name:             "aws",
+		stopIfSuccessful: false,
+		pf:               fromEC2,
+	},
+}
+
+// Get returns the cached hostname for the tracer, empty if we haven't found one yet.
+// Spawning a go routine to update the hostname if it is empty or out of date
+func Get() string {
+	now := time.Now()
+	var (
+		ch      string
+		expired bool
+		pv      string
+	)
+	// if provider is fargate never refresh
+	// Otherwise, refresh on expiration or if hostname hasn't been found.
+	if ch, pv, expired = getCached(now); pv == fargateName || (!expired && ch != "") {
+		return ch
+	}
+	// Use CAS to avoid spawning more than one go-routine trying to update the cached hostname
+	ir := isRefreshing.CompareAndSwap(false, true)
+	if ir {
+		go func() {
+			updateHostname(now)
+		}()
+	}
+	return ch
+}
+
+func updateHostname(now time.Time) {
+	defer isRefreshing.Store(false)
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+	var hostname string
+	var hnProvider string
+
+	for _, p := range providerCatalog {
+		detectedHostname, err := p.pf(ctx, hostname)
+		if err != nil {
+			log.Debug("Unable to get hostname from provider %s: %v", p.name, err)
+			continue
+		}
+		hostname = detectedHostname
+		hnProvider = p.name
+		log.Debug("Found hostname %s, from provider %s", hostname, p.name)
+		if p.stopIfSuccessful {
+			log.Debug("Hostname detection stopping early")
+			setCached(now, hostname, p.name)
+			return
+		}
+	}
+	if hostname != "" {
```

Comment:
If I understand correctly, an empty string is a valid value when running on Fargate. Are we ignoring it on purpose here? (not sending or caching or sending empty hostnames even on Fargate)